straxen/plugins/peak_processing.py

```diff
@@ -331,98 +336,393 @@ def find_n_competing(peaks, windows, fraction):
+                                  is_position, 
+                                  self.getsigma(self.config['shadow_sigma_and_baseline'], current_peak['area']))
+            
+            # Fill results
+            names = ['shadow', 'dt']
+            if 's2' in key: # Only previous S2 peaks have (x,y)
+                names += ['x', 'y']
+            if 'time' in key: # Only time shadow gives the nearest large peak
+                names += ['nearest_dt']
+            for name in names:
+                if name == 'nearest_dt':
+                    result[f'{name}_{type_str}'] = array[name]
+                else:
+                    result[f'{name}_{key}'] = array[name]
+
+        distance = np.sqrt((result[f'x_s2_position_shadow'] - current_peak['x']) ** 2 + 
+                           (result[f'y_s2_position_shadow'] - current_peak['y']) ** 2)
+        # If distance is NaN, set largest distance
+        distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+        # HalfCauchy PDF when calculating S2 position shadow
+        result['pdf_s2_position_shadow'] = halfcauchy.pdf(distance, 
+                                                          scale=self.getsigma(self.config['shadow_sigma_and_baseline'], current_peak['area']))
+
+        # 6. Set time and endtime for peaks
+        result['time'] = current_peak['time']
+        result['endtime'] = strax.endtime(current_peak)
+        return result
+
+    @staticmethod
+    def getsigma(sigma_and_baseline, s2):
+        # The parameter of HalfCauchy, which is a function of S2 area
+        return sigma_and_baseline[0] / np.sqrt(s2) + sigma_and_baseline[1]
 
     @staticmethod
     @numba.njit
-    def compute_shadow(peaks, pre_s2_peaks, touching_windows, exponent, res):
+    def peaks_shadow(peaks, pre_peaks, 
+                     touching_windows, exponent, 
+                     result, pos_corr, sigmas=None):
         """
-        For each peak in peaks, check if there is a shadow-casting S2 peak
+        For each peak in peaks, check if there is a shadow-casting peak
         and check if it casts the largest shadow
         """
-        for p_i, p_a in enumerate(peaks):
-            # reset for every peak
-            new_shadow = 0
-            s2_indices = touching_windows[p_i]
-            for s2_idx in range(s2_indices[0], s2_indices[1]):
-                s2_a = pre_s2_peaks[s2_idx]
-                if p_a['center_time'] - s2_a['center_time'] <= 0:
+        for p_i, (suspicious_peak, sigma) in enumerate(zip(peaks, sigmas)):
+            # casting_peak is the previous large peak casting shadow
+            # suspicious_peak is the suspicious peak which in shadow from casting_peak
+            indices = touching_windows[p_i]
+            for idx in range(indices[0], indices[1]):
+                casting_peak = pre_peaks[idx]
+                dt = suspicious_peak['center_time'] - casting_peak['center_time']
+                if dt <= 0:
+                    continue
```

Comment:
I think you can even `break` here is not it?. Peaks should be sorted by time. Although I think the difference is small.